pkg/resources/validation/podsecurityadmissionconfigurationtemplate/podsecurityact.go

```diff
@@ -0,0 +1,295 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/pod-security-admission/api"
+	"k8s.io/utils/trace"
+)
+
+var GVR = schema.GroupVersionResource{
+	Group:    "management.cattle.io",
+	Version:  "v3",
+	Resource: "podsecurityadmissionconfigurationtemplates",
+}
+
+type Validator struct {
+	ManagementClusterCache   v3.ClusterCache
+	provisioningClusterCache v1.ClusterCache
+}
+
+const ByPodSecurityAdmissionConfigurationName = "podSecurityAdmissionConfigurationName"
+
+func NewPodSecurityAdmissionConfigurationTemplateValidator(clients *clients.Clients) *Validator {
+	return &Validator{
+		ManagementClusterCache:   clients.Management.Cluster().Cache(),
+		provisioningClusterCache: clients.Provisioning.Cluster().Cache(),
+	}
+}
+
+func (v *Validator) ValidatingWebhook(clientConfig admissionregistrationv1.WebhookClientConfig) *admissionregistrationv1.ValidatingWebhook {
+	valWebhook := admission.NewDefaultValidatingWebhook(v, clientConfig, admissionregistrationv1.AllScopes)
+	valWebhook.FailurePolicy = admission.Ptr(admissionregistrationv1.Ignore)
+
+	v.ManagementClusterCache.AddIndexer(ByPodSecurityAdmissionConfigurationName, ByPodSecurityAdmissionConfigurationTemplateV3)
+	v.provisioningClusterCache.AddIndexer(ByPodSecurityAdmissionConfigurationName, ByPodSecurityAdmissionConfigurationTemplateV1)
+
+	return valWebhook
+}
+
+func ByPodSecurityAdmissionConfigurationTemplateV1(obj *provv1.Cluster) ([]string, error) {
+	if obj.Spec.DefaultPodSecurityAdmissionConfigurationTemplateName == "" {
+		return nil, nil
+	}
+	return []string{obj.Spec.DefaultPodSecurityAdmissionConfigurationTemplateName}, nil
+}
+
+func ByPodSecurityAdmissionConfigurationTemplateV3(obj *mgmtv3.Cluster) ([]string, error) {
+	if obj.Spec.DefaultPodSecurityAdmissionConfigurationTemplateName == "" {
+		return nil, nil
+	}
+	return []string{obj.Spec.DefaultPodSecurityAdmissionConfigurationTemplateName}, nil
+}
+
+func (v *Validator) GVR() schema.GroupVersionResource {
+	return GVR
+}
+
+func (v *Validator) Operations() []admissionregistrationv1.OperationType {
+	return []admissionregistrationv1.OperationType{admissionregistrationv1.Update, admissionregistrationv1.Create, admissionregistrationv1.Delete}
+}
+
+func (v *Validator) Admit(req *admission.Request) (*admissionv1.AdmissionResponse, error) {
+	listTrace := trace.New("PodSecurityAdmissionConfigurationTemplate Admit", trace.Field{Key: "user", Value: req.UserInfo.Username})
+	defer listTrace.LogIfLong(2 * time.Second)
+	resp := &admissionv1.AdmissionResponse{}
+	oldTemplate, newTemplate, err := objectsv3.PodSecurityAdmissionConfigurationTemplateOldAndNewFromRequest(&req.AdmissionRequest)
+	if err != nil {
+		return resp, err
+	}
+
+	if req.Operation == admissionv1.Create || req.Operation == admissionv1.Update {
+		v.ValidateConfiguration(newTemplate, resp)
+		if resp.Result != nil {
+			return resp, nil
+		}
```

Comment:
It is not a great paradigm to pass in a variable `response` to be mutated and then checked for specific mutations. `ValidateConfiguration` should just return an error if any and then the response can be created with that error and returned. 

```suggestion
		errList := v.ValidateConfiguration(newTemplate)
		if errList.ToAggregate() != nil {
			return &admissionv1.AdmissionResponse{
				Result: &metav1.Status{
					Status:  "Failure",
					Message: errList.ToAggregate().Error(),
					Reason:  metav1.StatusReasonBadRequest,
					Code:    http.StatusUnprocessableEntity,
				},
			}, nil
		}
```

An alternative approach if you do not want the `ErrorList` logic this high up in `Admit(...)` is to rename `v.ValidateConfiguration(newTemplate)` to `handleCreateAndUpdate` and just return the response and and error.
```suggestion
		return v.handleCreateAndUpdate(newTemplate)
```